Feat: Upload user's post

bbianca2004 opened the pull request:
#  Refactor Outfit Posting Flow

## Summary
Refactored the outfit posting flow to simplify data handling, eliminate premature Firestore writes, and centralize post management.  
Users now:
- Select their photo and description in **FitCheck**
- Preview and manage the post in **Overview**
- Add linked items in **AddItem**
- Publish once everything is ready

Only the final published post (and its items) are saved to Firestore.

---

## What’s Changed

### UI
- **FitCheck Screen**
  - Collects outfit photo and description without saving to Firestore.
  - Passes image and description to Overview via navigation arguments.
- **Overview Screen**
  - Generates a new `postUuid` when opened.
  - Displays the selected image and description.
  - Loads and lists items linked to the current post.
  - Handles publishing (uploads photo, creates post in Firestore).
- **AddItem Screen**
  - Integrated with `postUuid` so new items automatically associate with the correct post.

###  ViewModel
- Refactored `OutfitOverviewViewModel`:
  - Generates and manages the `postUuid`.
  - Loads and filters items by `postUuid`.
  - Handles photo uploads and Firestore post creation.
  - Provides success/error state management.
- Removed early Firestore writes from FitCheck.
- Simplified data flow between **FitCheck → Overview → AddItem**.

###  Documentation
- Updated KDoc for the new `OutfitOverviewViewModel` and  `FitCheck` and refactored posting flow.

---

## Why
This refactor:
- Prevents leftover Firestore writes when users change or cancel a post.
- Centralizes all post logic in the Overview screen (“draft hub”).
- Keeps Firestore clean, only published posts are saved.
- Reduces race conditions and redundant uploads.
- Simplifies debugging and improves maintainability.



---

## Verification Checklist
- [x] Verified navigation: **FitCheck → Overview → AddItem → Publish**
- [x] Confirmed consistent `postUuid` across screens
- [x] Ensured only published posts appear in Firestore
- [x] Verified cancelling deletes draft items
- [x] Added OutfitPreview, Fitcheck View model and Screen tests

> [!WARNING]  
> This PR exceeds the 1.5k lines limit, but I think it was necessary to do it all in one PR since the screens were already implemented, however it was a lot of heavy refactoring of the view models regarding the navigation arguments they pass between themselves. Splitting it would have generated some not so meaningfull PR's that could not be properly tested.

---
Closes #8

cocoStein commented on app/src/main/java/com/android/ootd/ui/post/FitCheckViewModel.kt:
```diff
@@ -2,24 +2,45 @@ package com.android.ootd.ui.post
+        itemsRepo.deletePostItems(postUuid)
+      } catch (e: Exception) {
+        // optional: log or show error
+        e.printStackTrace()
```
Maybe instead of printing the whole stack you could just print a log error:

```suggestion
        Log.e("FitCheckViewModel", "Failed to delete items for post: $postUuid", e)
```

I think it makes more sense and follow what we did in the rest of the project, lmkwyt ?

cocoStein commented on app/src/main/java/com/android/ootd/ui/post/OutfitPreviewViewModel.kt:
```diff
@@ -33,50 +41,145 @@ data class PreviewUIState(
-        _uiState.value = PreviewUIState(items = items, isLoading = false)
+        val user = Firebase.auth.currentUser ?: throw Exception("User not logged in")
+
+        Log.d("OutfitPreviewViewModel", "Publishing post: ${state.postUuid}")
```
I know this is a draft but don't forget to remove the `log.d` for the final version of this PR 🔥

cocoStein commented on app/src/main/java/com/android/ootd/ui/post/OutfitPreviewViewModel.kt:
```diff
@@ -33,50 +41,145 @@ data class PreviewUIState(
     }
   }
+
+  //  fun refreshItems() {
```
Same with the commented out code

cocoStein commented on app/src/main/java/com/android/ootd/ui/post/OutfitPreviewViewModel.kt:
```diff
@@ -33,50 +41,145 @@ data class PreviewUIState(
+    _uiState.value = _uiState.value.copy(errorMessage = message)
+  }
+
+  fun clearSuccessMessage() {
```
Same thing as above, don't forget to remove this as it's currently unused.

I think you forgot to call this function after you successfully posted !

cocoStein commented on app/src/main/java/com/android/ootd/ui/post/PreviewItemScreen.kt:
```diff
@@ -82,25 +83,41 @@ object PreviewItemScreenTestTags {
+        onPostSuccess()
+      }
+    }
+  }
```
This is a really good idea. For now, you added a simple toast message, but I feel like later on we could add a full-screen animation to let the user know he has posted !!

bbianca2004 replied:
Yes, I totally agree with that, for the sake of time I left it out for the moment, but in the future we could definitely add it simmilarly to what you did in AddItem and EditItem screen.

cocoStein reviewed (commented):
This is a really good draft, you’re almost there! 
I see that you really understand the architecture of the project and the firebase as you handled this complex task without making much mistakes. The navigation you added is also looking great.

As we discussed, maybe investigate the bug of the posts and I added some small comments, but it looks all good to me

bbianca2004 commented on app/src/main/java/com/android/ootd/ui/post/OutfitPreviewViewModel.kt:
```diff
@@ -17,11 +22,14 @@ import kotlinx.coroutines.launch
     val errorMessage: String? = null,
+    val successMessage: String? = null,
     val isLoading: Boolean = false,
+    val isPublished: Boolean = false
```
Thank youu!

bbianca2004 commented on app/src/main/java/com/android/ootd/model/items/ItemsRepositoryFirestore.kt:
```diff
@@ -26,6 +26,21 @@ class ItemsRepositoryFirestore(private val db: FirebaseFirestore) : ItemsReposit
+    val snapshot =
+        db.collection(ITEMS_COLLECTION)
+            .whereEqualTo("postUuid", postUuid)
+            .whereEqualTo("ownerId", ownerId)
```
Yes you are right, I removed the hardcoded strings and used the constant, thanks for spotting this :))

bbianca2004 commented on app/src/main/java/com/android/ootd/model/items/ItemsRepositoryFirestore.kt:
```diff
@@ -26,6 +26,21 @@ class ItemsRepositoryFirestore(private val db: FirebaseFirestore) : ItemsReposit
+
+    val snapshot =
+        db.collection(ITEMS_COLLECTION)
+            .whereEqualTo("postUuid", postUuid)
```
Yess, good suggestion, will do it!

bbianca2004 commented:
Thank you to the two of you for your comments, they were very good observations and I implemented all the fixes you suggested. I also added the circular indicator, you two had such a good idea when implementing this in your screens, now the user can see that their request is being processed and not have to be confused by what is happening when the post is being created.

asunluer commented on app/src/androidTest/java/com/android/ootd/ui/post/PreviewItemScreenTest.kt:
```diff
@@ -364,33 +413,50 @@ class PreviewItemScreenTest : ItemsTest by InMemoryItem {
+  //    // After the failure, we expect:
+  //    assert(viewModel.uiState.value.items.isEmpty())
+  //    assert(viewModel.uiState.value.errorMessage?.contains("Failed to load items") == true)
+  //  }
```
If you will not use this test, please don't forget to remove it. :)

asunluer commented on app/src/androidTest/java/com/android/ootd/ui/post/PreviewItemScreenTest.kt:
```diff
@@ -651,75 +695,97 @@ class PreviewItemScreenTest : ItemsTest by InMemoryItem {
+  //      //  Depending on timing, items might still contain old data.
+  //      // To make test robust, just assert that it’s *not newly loaded*.
+  //      assert(finalState.items.isEmpty() || finalState.items == firstState.items)
+  //  }
```
Same as above.

cocoStein reviewed (approved):
Hello, first of all great job managing to complete this task, it was one of the hardest 💪🏼

As Asli said, just remove the commented out code, and for my part i just want to ask if you can try to remove the 2 new sonar issue. They are not blocking issue so don't spend to much time on them, but if you can do it it would be nice 🔥
